pkg/controller/bundlec/deployment.go

```diff
@@ -0,0 +1,154 @@
+package bundlec
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"sort"
+
+	"github.com/atlassian/smith"
+	"github.com/atlassian/smith/pkg/util"
+	"github.com/pkg/errors"
+	apps_v1 "k8s.io/api/apps/v1"
+	core_v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+const (
+	envFromChecksumAnnotation = smith.Domain + "/envFromChecksum"
+)
+
+// works around https://github.com/kubernetes/kubernetes/issues/22368
+func (st *resourceSyncTask) forceDeploymentUpdates(spec *unstructured.Unstructured, actual runtime.Object, namespace string) (specRet *unstructured.Unstructured, e error) {
+	gvk := spec.GroupVersionKind()
+
+	if gvk.Group == apps_v1.GroupName && gvk.Kind == "Deployment" {
+		return st.processDeployment(spec, actual, namespace)
+	}
+
+	return spec, nil
+}
+
+func (st *resourceSyncTask) processDeployment(spec *unstructured.Unstructured, actual runtime.Object, namespace string) (specRet *unstructured.Unstructured, e error) {
+	deploymentSpec := &apps_v1.Deployment{}
+
+	if err := util.ConvertType(st.scheme, spec, deploymentSpec); err != nil {
+		return nil, errors.Wrap(err, "failure to parse Deployment")
+	}
+
+	hasEnvFrom := false
+	for _, container := range deploymentSpec.Spec.Template.Spec.Containers {
+		if len(container.EnvFrom) != 0 {
+			hasEnvFrom = true
+			break
+		}
+	}
+	if !hasEnvFrom {
```

Comment:
This is not right - if there was a secret/configmap reference and it was removed, in the end the object will still have the annotation with the same value and will not be reprocessed by the deployment controller. But we want it to be reprocessed because the ref was removed, right?
We must change the value of the annotation here (or remove the annotation, but that is hard to do given current spec check implementation which merges annotations).

Reply:
@ash2k I agree that the annotation needs to be deleted, but I don't think leaving annotation means that Deployment won't be reprocessed. To remove envFrom reference you have to modify pod template spec, therefore replica set will be re-created with a normal logic, no annotation hack is required in that case.

Reply:
You are right, I didn't consider this case carefully enough, sorry.
I still think it's better to do something about the annotation - leaving it with previous value is misleading. Given that we cannot easily delete it... lets give it a different value. Empty string or perhaps hash of an empty slice to keep it uniform.

Reply:
> Given that we cannot easily delete it

Hang on, so I can't just delete(..., myKey)?

Reply:
No, because spec and actual object's annotations are merged. Yes, there is no way to remove an annotation via a Bundle - we need server-side apply. See
https://github.com/atlassian/smith/blob/master/pkg/speccheck/speccheck.go#L66
https://github.com/atlassian/smith/blob/master/pkg/speccheck/speccheck.go#L150-L163